### DISCUSSION TITLE
Using batch mode of tau-argus for protecting already existing tables

### DISCUSSION THREAD

**Discussion by Benjamin-Fuchs** [sdcTools/tauargus#10] [Q&A] [answered]
Dear all, 

I try to write an executable batch file for protecting already existing tables. At the moment I test all the possibilities that tau-argus offer. I have several questions, I would like to ask. I hope someone can help me. I use tau-argus version 4.2.3 build 1. 
Q1:  I define a response variable, in my case it has the name “Fallzahlen”, as frequency. When using the name of the response variable in the command <SPECIFYTABLE> (e.g.: <SPECIFYTABLE> “ExpVar1” “ExpVar2” | “Fallzahlen” | “” | “”) I get the following error message. When using the term “<freq>” instead of the actual name of the response variable the command works. Does anyone know what the reason for this error message is?

![errormessage01](https://github.com/sdcTools/tauargus/assets/126875531/0cea00fb-1969-47e7-a7c2-93fe49dcd9ac)

For explanation: My table includes the number of different types of vehicles for different German federal states. The first line of my input file includes “Fahrzeuge insgesamt”, which is the code for the total of my first explanatory variable vehicle (in German: Fahrzeuge) and Baden-Wuerttemberg is a German federal state. The third column shows the corresponding number of cases, here: 9 638 138.

Q2: According to the manual 4.1, p.122 one can define lambda on the fifth position of the command <SPECIFYTABLE> with 1 as default value. I tried to define another value for lambda but always received the following error message (in the error message I tried to define lambda = 0). Under the assumption that I haven’t done anything wrong, I assume that SPECIFYTABLE interprets my input for lambda as part of the cost variable one can define on the fourth position of the command. Is this a bug or have I done something wrong? When I do not set lambda actively and only define the first four positions (explanatory variable, response variable, shadow variable and cost variable) the command works.

![errormessage02](https://github.com/sdcTools/tauargus/assets/126875531/a7a72a06-cc4a-4ecd-823d-cc035ba112d4)

Q3: In manual 4.1 on p.123 the option MOD of the command <SUPPRESS> allows to define a maximum number of time per subtable. Just out of sheer curiosity: What happens if tau-argus need longer than the defined maximum time?

Q4: At the moment I am able to implement a minimum frequency rule. Whenever I try to implement a minimum frequency rule and a p%-rule at the same time, I get the following error message.

![errormessage03](https://github.com/sdcTools/tauargus/assets/126875531/3e2721ae-f531-4f3e-89a1-6e6838baaea7)

Alternatively, I tried to implement a minimum frequency rule and a p%-rule manually via the menu. When I define the response variable as numeric, then firstly, I am not able to set a minimum frequency rule anymore, and secondly, after defining a p-rule (e.g. P = 10 and n = 1) I get the following error message. 

![errormessage04](https://github.com/sdcTools/tauargus/assets/126875531/fde19d37-5ffd-437f-bac6-672fe0a18ec3)

I get the same error message when defining the response variable as frequency. In this case, however I am able to define a minimum frequency rule in addition to a p%-rule at least.

Any help is greatly appreciated.

**Comment by sven-grunwald** [comment, marked as answer]
Hi Benjamin,

first of all I think you will run into several problems when protecting frequency tables with TauArgus (instead of magnitude tables). At least that are our experiences.

Q1) The call should look like “ExpVar1” “ExpVar2” | “Fallzahlen” | “Fallzahlen” | “Fallzahlen” in your case. The second "Fallzahlen" is the magnitude value (so its the same as your frequency if you don't have magnitude tables) and the third "Fallzahlen" describes the cost fiúnction. TauArgus tries to optimize the secondary suppression pattern according to the cost function (to reduce the costs). If we assume that you have more loss of information if you have larger frequencys, then you can just take your frequencys as cost variable.

Q2) I think the value for lambda should be larger than 0. Tha lambda value specifies how the costs are calculated. The calculated costs should look like cost^lambda, so if you choose lambda = 0 then every cell would have the same costs of 1 after transformation. Maybe that doesn't make much sense. The threshold is lambda = 1 (everything stays just as you defined it). If you want to increase your probability that larger values get suppressed instead of smaller, you can choose a value between 0 and 1 (e.g. 0.2 or 0.8).

Q3) The time you can specifiy in the MOD statement is just the time that is spent for every single subtable to be optimised. So if in the specified time there cannot be found the optimal solution, there should be at least a non-optimal solution for your problem. As far as the time limit you specify is just for subtables, the calculation time for a whole table can easily succeed that limit in case of executing hierarchical tables.

Q4) Sorry, I have no answer for this since we don't aggregate our tables inside of TauArgus. 

Best,
Sven

**Comment by Benjamin-Fuchs** [comment]
Hi Sven,

thank you very much for your quick and detailed response!

In the end I want to protect magnitude tables. But I thought I start with protecting frequency tables as a first step to learn how to write a batch file in tau-argus. I was successful in implementing a minimum frequency rule at least. 

The next step is to protect magnitude tables with a minimum frequency rule and a p%-rule combined. I have forgotten to mention that Q1/2 corresponds to an input data file with 2 explanatory variables and frequency as response variable. Q4/5 corresponds to an input data file with 2 explanatory variables, the magnitude as response variable, and 2 variables with frequencies of the 2 largest contributors.
To your answer to Q2: I tried several values between 0 and 1. Unfortunately, I always get the same error message (instead of 0 the error message shows another value I used for lambda, e.g. 0.2 or 0.3). 

Best regards,
Benjamin

**Reply by sven-grunwald** [reply]
Hi Benjamin,

I think the p% rule doesn't make sense at all for frequency tables. Your single values would always be 1 (since this is the frequency of each unit), so the largest and the second largest contributer would be equal 1 and so the result would be the same as you take frequency rule with suppressions for lower than 3 contributers.

Concerning Q1: Already understood that. I just wanted to show that if you want to specify a frequency table, the call looks like “ExpVar1” “ExpVar2” | “Fallzahlen” | “Fallzahlen” | “Fallzahlen”, so your (not existing) magnitude value and the cost value is replaced by your frequency variable. But Tau Argus needs variables at this place to compute the costs for the optimization.

Concerning Q2: The lambda value is written without the qoute sign, so just write “ExpVar1” “ExpVar2” | “Fallzahlen” | “Fallzahlen” | “Fallzahlen”,0.2. Did you try this?

**Reply by Benjamin-Fuchs** [reply]
Hi Sven,

thanks a lot for your reply. 
Concerning Q2:  With or without quote signs, I still get the same error message. 

Concerning the other points: I think we talk past each other (at least partially). I would stop this thread at this point. Thank you very much for your help I've got a perfect answer for Q3. Q1 and Q2 I think we talk past each other and Q4 is my main concern at the moment.

Best regards
Benjamin

**Comment by ppdewolf** [comment]
Hi Benjamin, 

Concerning your remark `Q4/5 corresponds to an input data file with 2 explanatory variables, the magnitude as response variable, and 2 variables with frequencies of the 2 largest contributors.`
Could you provide the `.rda` file? That would make it easier to assess the situation.
I understand that in this case your input is still tabular data, but with additional information for each table cell?

Moreover, what do you mean by `frequencies of the 2 largest contributors`? Is it a magnitude variable representing integers (e.g. number of employees)? Because "frequency tables" in tau-argus are tables counting the number of contributors, not adding frequencies (like number of employees) of each contributor. 

When there is no variable with the tag `<FREQUENCY>` in the `.rda` file, there will be no option to use the threshold rule in the interface, since then frequency information is not available.

Note:
According to the manual, FREQ(80,5) would mean a minimum number of 80 and a 5% safety range. I assume you would want this the other way around, i.e., FREQ(5,80)?

Best,
Peter-Paul

**Comment by Benjamin-Fuchs** [comment]
Hi Peter-Paul,

as already mentioned in the other thread, your explanation solved my problem. I just want to respond to your note:

FREQ(80,5) was intended. In the file that I used to test tau-argus, the lowest magnitude in a cell was 76. To see some results after applying safety rules I defined the minimum frequency rule to block all cells with less than 80 vehicles.

Again, thank you very much for your help!

Best,
Benjamin
